### DISCUSSION TITLE
The git plugin has too many aliases

### DISCUSSION THREAD

**Discussion by mcornella** [ohmyzsh/ohmyzsh#11969] [Ideas]
This is a recurring sentiment:

https://twitter.com/RaymondGulbr/status/808964779121475585:
![screenshot of Raymond Gulbrandsen's tweet](https://cloud.githubusercontent.com/assets/1441704/21396621/2713d72e-c7a1-11e6-8504-f476a6519d20.png)

https://github.com/robbyrussell/oh-my-zsh/issues/3288#issuecomment-145481037
https://github.com/robbyrussell/oh-my-zsh/issues/3288#issuecomment-267463829
https://github.com/robbyrussell/oh-my-zsh/issues/3682#issuecomment-78437079

This has many harmful side-effects, mainly added commands (some of which are destructive and at one keystroke). Some aliases also might override installed commands, which then need to be selectively disabled or renamed. There is #4585 pending but no closer to being merged.

Meanwhile, more and more people want to add their own little alias, without thinking about the added benefits or costs to the common interest.

One possible fix for that is having bundles of aliases for specific workflows, such that if certain git commands are used together very frequently we alias them and group these aliases into specific functions or if branches. For example, if one does not use commit signing regularly, not having that be there by default.

This is a discussion issue and it is in a very early stage, so we're quite far from a PR fixing this.
What other solutions can you think of?

**Comment by obcilion** [comment]
I'd be fine with removing all but the most common ones (add, status, branch etc.) and letting users define their own after that. Having bundles is also a good idea, but how would installing of alias bundles be managed?

**Comment by mcornella** [comment]
My thinking is maybe adding some configuration variables, like `$GIT_PLUGIN_USE_SIGNING`, and having one for each bundle of aliases. The user would then set up these configuration variables in their zshrc file.

Another option would be to have multiple functions that define such aliases, and again, the user would call them in their zshrc files. 

I'm open to other suggestions and I'll think about other options we may have. But yes, many aliases could be purged from the main plugin and leave only the absolute necessary.

**Comment by muesli** [comment]
I was getting rather annoyed by all kinds of "random" aliases defined on my system by various oh-my-zsh plugins, so I went ahead and separated the aliases from the plugins for the git and golang plugin at least.

See here:
https://github.com/muesli/oh-my-zsh/commit/9db77ac56f01458ea8aad30d75aa2a885cd5c345

This allows you to load aliases separately in your .zshrc:
aliasdefs=(golang)

**Comment by shinenelson** [comment]
I'd prefer just segregating them into multiple plugins based on their usage. (Just my 2 ¢)

**Comment by noraj** [comment]
`gm` alias is problematic when using **Graphics Magick**, this is the name of the main command, see http://stackoverflow.com/questions/29452564/git-fatal-error-when-using-graphics-magick-convert-command. I also said that [here](https://github.com/robbyrussell/oh-my-zsh/issues/6832#issuecomment-390722815).

Any solution to remove those useless aliases but keeping plugins ?

**Comment by noraj** [comment]
As I said in #5115. The solution is easy. oh-my-zsh should provide a switch to disable the aliases. No need to remove them.

**Comment by iMouath** [comment]
There's also some dangerous commands that should **not** be aliased.
example:

```
**gpristine='git reset --hard && git clean -dfx'**
grh='git reset'
grhh='git reset --hard'
groh='git reset origin/$(git_current_branch) --hard'
gunwip='git log -n 1 | grep -q -c "\-\-wip\-\-" && git reset HEAD~1'
```

**Comment by lgedeon** [comment]
Everybody wants their own list of aliases, but that hinders portability. I stopped using all aliases years ago because I kept being nearly disabled when I was trying to help someone else on their computer. I would spend more time trying to remember the verbose command than actually helping. And now there are dotfile repos, but before that, getting a new computer was a nightmare instead of something exciting because I had to find all the places that I had made modifications to my system.

It would be really nice (instead of hundreds of generic aliases across scores of plugins) if omz had a utility for quickly adding aliases to a dotfiles repo and a simple one-line command to switch to zsh, install (temporarily?) omz, and load aliases (and everything else) from your dotfile repo.

This could be implemented as a brew, a node package, and maybe one or two other small install scripts that accept your github/bitbucket repo name and it does the rest. If we do that, we could remove all the aliases and make moving to a new machine easier.

**Comment by ouuan** [comment]
I learned many git commands and found some better practice in the OMZ git aliases. For me, these aliases are like a collection of best practices.

**Comment by askpatrickw** [comment]
One of the problems I've had is how to learn the aliases?

Any thought about a `g?` to output the list with descriptions?
Would be great if you could do `gx?` and see aliases which start with `gx` but I'd be happy with just `g?` displaying all.
If that's interesting, I'll open a separate issue.

**Comment by ouuan** [comment]
> One of the problems I've had is how to learn the aliases?

I'm using [alias tips](https://github.com/djui/alias-tips) and `alias al='alias | G'` to search like `al commit`.

**Comment by askpatrickw** [comment]
Thank you @ouuan that plugin is awesome!
